core/instantiate.ml

```diff
@@ -329,7 +339,15 @@ let apply_type : Types.datatype -> Types.type_arg list -> Types.datatype =
                         mapstrcat ", " (fun t -> Types.string_of_type_arg t) tyargs))
         vars tyargs (IntMap.empty, IntMap.empty, IntMap.empty)
     in
-      instantiate_datatype (tenv, renv, penv) t
+    if remaining_quantifiers = [] then
```

Comment:
I think it is cleaner to use pattern-matching instead of a bare if-then-else here.